apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/Keys.cs

```diff
@@ -275,49 +302,72 @@ private V1.KeyRange ToProtobuf()
     /// </summary>
     public sealed class Key
     {
-        /// <summary>
-        /// The values of this key.
-        /// </summary>
-        internal ListValue KeyParts { get; }
+        internal SpannerParameterCollection Parameters { get; }
 
         /// <summary>
         /// Creates a key from the values in the parameter collection.
         /// </summary>
         /// <param name="keyParts">The values that define the key</param>
         internal Key(SpannerParameterCollection keyParts)
         {
-            KeyParts = ToListValue(keyParts);
+            Parameters = keyParts.Clone();
         }
 
+        // Note: We could have just used IEnumerable<SpannerParameter>, but this constructor overload accepting params object[]
+        // gets confused, so explicitly creating new SpannerParameterCollection to disambiguate. 
+
         /// <summary>
         /// Creates a key consisting of the given key parts.
         /// </summary>
         /// <param name="keyParts">The values that define the key</param>
-        public Key(params object[] keyParts)
+        public Key(params object[] keyParts) : this(new SpannerParameterCollection(keyParts.Select(j => new SpannerParameter { Value = j })))
         {
-            var protobufValues = keyParts.Select(
-                v => SpannerDbType.FromClrType(v?.GetType())?.ToProtobufValue(v) ?? Value.ForNull());
-            KeyParts = new ListValue { Values = { protobufValues } };
         }
 
+        /// <summary>
+        /// Gets the values of this Key.
+        /// </summary>
+        /// <param name="options">The options for conversion between protobuf and CLR types.</param>
+        /// <returns>The values of this Key.</returns>
+        internal ListValue GetKeyParts(SpannerConversionOptions options) => ToListValue(Parameters, options);
+
         /// <inheritdoc/>
         public override bool Equals(object other)
-            => other is Key key && KeyParts.Equals(key.KeyParts);
+        {
+            if (other is Key key)
+            {
+                if (Parameters == null && key.Parameters == null)
+                {
+                    return true;
+                }
+                else if (Parameters == null || key.Parameters == null)
+                {
+                    return false;
+                }
+                else
+                {
+                    return Parameters.Equals(key.Parameters);
+                }
+            }
+
+            return false;
+        }
 
         /// <inheritdoc/>
-        public override int GetHashCode() => KeyParts.GetHashCode();
+        public override int GetHashCode() => Parameters.GetHashCode();
 
+        // This is to make the implementation backward compatible.
         /// <inheritdoc/>
-        public override string ToString() => KeyParts.ToString();
+        public override string ToString() => ToListValue(Parameters, SpannerConversionOptions.Default).ToString();
```

Comment:
This will be a little confusing as the string representation will always use the default conversion options. Maybe you should consider adding a `ToString(SpannerConversionOptions options)` method as well.

Another option could be to include the `SpannerConversionOptions` as a property of this class, so you don't have to add it as an argument to all the methods in the class.

Reply:
Unfortunately `SpannerConversionOptions` is `internal` so we can't add a public overload of `ToString` that takes `SpannerConversionOptions` as a parameter.

I can add a property, but it will end up having `SpannerConversionOptions.Default` as its value or am I missing something?